File: code/06-search-trees/python/btree.py
"""BTree: distinct int key, CLRS minimum degree t, top-down split-before-
descend insert, borrow/merge delete. lecture-notes/code/lecture06/java/
BTree.java와 같은 정책·예제 입력(t=2)을 쓴다."""


# snippet:btree:start
class _Node:
    def __init__(self, leaf):
        self.keys = []
        self.children = []
        self.leaf = leaf


def _lower_bound(a, k):
    i = 0
    while i < len(a) and a[i] < k:
        i += 1
    return i


class BTree:
    def __init__(self, minimum_degree):
        if minimum_degree < 2:
            raise ValueError("t >= 2")
        self.t = minimum_degree
        self.root = _Node(True)
        self._size = 0

    def __len__(self):
        return self._size

    def contains(self, k):
        return self._contains(self.root, k)

    def _contains(self, x, k):
        i = _lower_bound(x.keys, k)
        if i < len(x.keys) and x.keys[i] == k:
            return True
        return (not x.leaf) and self._contains(x.children[i], k)

    def insert(self, k):
        if self.contains(k):
            return False
        t = self.t
        if len(self.root.keys) == 2 * t - 1:
            s = _Node(False)
            s.children.append(self.root)
            self._split_child(s, 0)
            self.root = s
        self._insert_nonfull(self.root, k)
        self._size += 1
        return True

    def _insert_nonfull(self, x, k):
        i = _lower_bound(x.keys, k)
        if x.leaf:
            x.keys.insert(i, k)
            return
        t = self.t
        if len(x.children[i].keys) == 2 * t - 1:
            self._split_child(x, i)
            if k > x.keys[i]:
                i += 1
        self._insert_nonfull(x.children[i], k)

    def _split_child(self, x, i):
        t = self.t
        y = x.children[i]
        z = _Node(y.leaf)
        median = y.keys[t - 1]
        z.keys = y.keys[t:2 * t - 1]
        y.keys = y.keys[:t - 1]
        if not y.leaf:
            z.children = y.children[t:2 * t]
            y.children = y.children[:t]
        x.keys.insert(i, median)
        x.children.insert(i + 1, z)

    def delete(self, k):
        if not self.contains(k):
            return False
        self._delete(self.root, k)
        if not self.root.keys and not self.root.leaf:
            self.root = self.root.children[0]
        self._size -= 1
        return True

    def _delete(self, x, k):
        t = self.t
        i = _lower_bound(x.keys, k)
        if i < len(x.keys) and x.keys[i] == k:
            if x.leaf:
                x.keys.pop(i)
            elif len(x.children[i].keys) >= t:
                pred = self._max(x.children[i])
                x.keys[i] = pred
                self._delete(x.children[i], pred)
            elif len(x.children[i + 1].keys) >= t:
                succ = self._min(x.children[i + 1])
                x.keys[i] = succ
                self._delete(x.children[i + 1], succ)
            else:
                self._merge(x, i)
                self._delete(x.children[i], k)
            return
        child = i
        if len(x.children[child].keys) == t - 1:
            if child > 0 and len(x.children[child - 1].keys) >= t:
                self._borrow_prev(x, child)
            elif child < len(x.children) - 1 and len(x.children[child + 1].keys) >= t:
                self._borrow_next(x, child)
            else:
                if child < len(x.children) - 1:
                    self._merge(x, child)
                else:
                    self._merge(x, child - 1)
                    child -= 1
        self._delete(x.children[child], k)

    def _min(self, x):
        while not x.leaf:
            x = x.children[0]
        return x.keys[0]

    def _max(self, x):
        while not x.leaf:
            x = x.children[-1]
        return x.keys[-1]

    def _borrow_prev(self, x, i):
        c, s = x.children[i], x.children[i - 1]
        c.keys.insert(0, x.keys[i - 1])
        x.keys[i - 1] = s.keys.pop()
        if not c.leaf:
            c.children.insert(0, s.children.pop())

    def _borrow_next(self, x, i):
        c, s = x.children[i], x.children[i + 1]
        c.keys.append(x.keys[i])
        x.keys[i] = s.keys.pop(0)
        if not c.leaf:
            c.children.append(s.children.pop(0))

    def _merge(self, x, i):
        c, s = x.children[i], x.children.pop(i + 1)
        c.keys.append(x.keys.pop(i))
        c.keys.extend(s.keys)
        if not c.leaf:
            c.children.extend(s.children)

    def inorder(self):
        out = []

        def walk(x):
            for i, key in enumerate(x.keys):
                if not x.leaf:
                    walk(x.children[i])
                out.append(key)
            if not x.leaf:
                walk(x.children[len(x.keys)])

        walk(self.root)
        return out
# snippet:btree:end

if __name__ == "__main__":
    t = BTree(2)
    for k in [10, 20, 5, 6, 12, 30, 7, 17]:
        t.insert(k)
    print("inorder:", ",".join(str(k) for k in t.inorder()))
    print("size:", len(t))
    print("root:", ",".join(str(k) for k in t.root.keys))
    print("children:", ";".join(",".join(str(k) for k in c.keys) for c in t.root.children))
    for k in [6, 7, 5, 10, 12, 17, 20, 30]:
        t.delete(k)
    print("delete size:", len(t))

File: code/06-search-trees/c/btree.c
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>

// snippet:btree:start
/* BTree: distinct int key, CLRS minimum degree t, top-down split-before-
 * descend insert, borrow/merge delete.
 * lecture-notes/code/lecture06/java/BTree.java와 같은 정책·예제(t=2)를 C로 옮긴 것이다. */
#define MAX_KEYS 15 /* 2*t-1 for t up to 8, plenty for this chapter's t in {2,3,5} */
#define MAX_CHILDREN (MAX_KEYS + 1)

typedef struct Node {
    int keys[MAX_KEYS];
    int nkeys;
    struct Node *children[MAX_CHILDREN];
    bool leaf;
} Node;

typedef struct {
    int t;
    Node *root;
    int size;
} BTree;

static Node *node_new(bool leaf) {
    Node *n = malloc(sizeof *n);
    n->nkeys = 0;
    n->leaf = leaf;
    return n;
}

static void btree_init(BTree *b, int minimum_degree) {
    b->t = minimum_degree;
    b->root = node_new(true);
    b->size = 0;
}

static int lower_bound(const int *a, int n, int k) {
    int i = 0;
    while (i < n && a[i] < k) i++;
    return i;
}

static bool node_contains(const Node *x, int k) {
    int i = lower_bound(x->keys, x->nkeys, k);
    if (i < x->nkeys && x->keys[i] == k) return true;
    return !x->leaf && node_contains(x->children[i], k);
}

static bool btree_contains(BTree *b, int k) { return node_contains(b->root, k); }

static void split_child(BTree *b, Node *x, int i) {
    int t = b->t;
    Node *y = x->children[i];
    Node *z = node_new(y->leaf);
    int median = y->keys[t - 1];

    z->nkeys = t - 1;
    for (int j = 0; j < t - 1; j++) z->keys[j] = y->keys[t + j];
    if (!y->leaf) {
        for (int j = 0; j < t; j++) z->children[j] = y->children[t + j];
    }
    y->nkeys = t - 1;

    for (int j = x->nkeys; j > i; j--) x->keys[j] = x->keys[j - 1];
    for (int j = x->nkeys + 1; j > i + 1; j--) x->children[j] = x->children[j - 1];
    x->keys[i] = median;
    x->children[i + 1] = z;
    x->nkeys++;
}

static void insert_nonfull(BTree *b, Node *x, int k) {
    int i = lower_bound(x->keys, x->nkeys, k);
    if (x->leaf) {
        for (int j = x->nkeys; j > i; j--) x->keys[j] = x->keys[j - 1];
        x->keys[i] = k;
        x->nkeys++;
        return;
    }
    if (x->children[i]->nkeys == 2 * b->t - 1) {
        split_child(b, x, i);
        if (k > x->keys[i]) i++;
    }
    insert_nonfull(b, x->children[i], k);
}

static bool btree_insert(BTree *b, int k) {
    if (btree_contains(b, k)) return false;
    if (b->root->nkeys == 2 * b->t - 1) {
        Node *s = node_new(false);
        s->children[0] = b->root;
        b->root = s;
        split_child(b, s, 0);
    }
    insert_nonfull(b, b->root, k);
    b->size++;
    return true;
}

static int subtree_min(Node *x) {
    while (!x->leaf) x = x->children[0];
    return x->keys[0];
}

static int subtree_max(Node *x) {
    while (!x->leaf) x = x->children[x->nkeys];
    return x->keys[x->nkeys - 1];
}

static void merge_children(BTree *b, Node *x, int i) {
    Node *c = x->children[i], *s = x->children[i + 1];
    c->keys[c->nkeys] = x->keys[i];
    for (int j = 0; j < s->nkeys; j++) c->keys[c->nkeys + 1 + j] = s->keys[j];
    if (!c->leaf) {
        for (int j = 0; j <= s->nkeys; j++) c->children[c->nkeys + 1 + j] = s->children[j];
    }
    c->nkeys += 1 + s->nkeys;
    for (int j = i; j < x->nkeys - 1; j++) x->keys[j] = x->keys[j + 1];
    for (int j = i + 1; j < x->nkeys; j++) x->children[j] = x->children[j + 1];
    x->nkeys--;
    free(s);
    (void)b;
}

static void borrow_prev(Node *x, int i) {
    Node *c = x->children[i], *s = x->children[i - 1];
    for (int j = c->nkeys; j > 0; j--) c->keys[j] = c->keys[j - 1];
    c->keys[0] = x->keys[i - 1];
    x->keys[i - 1] = s->keys[s->nkeys - 1];
    if (!c->leaf) {
        for (int j = c->nkeys + 1; j > 0; j--) c->children[j] = c->children[j - 1];
        c->children[0] = s->children[s->nkeys];
    }
    c->nkeys++;
    s->nkeys--;
}

static void borrow_next(Node *x, int i) {
    Node *c = x->children[i], *s = x->children[i + 1];
    c->keys[c->nkeys] = x->keys[i];
    x->keys[i] = s->keys[0];
    if (!c->leaf) c->children[c->nkeys + 1] = s->children[0];
    c->nkeys++;
    for (int j = 0; j < s->nkeys - 1; j++) s->keys[j] = s->keys[j + 1];
    if (!s->leaf) {
        for (int j = 0; j < s->nkeys; j++) s->children[j] = s->children[j + 1];
    }
    s->nkeys--;
}

static void btree_delete_node(BTree *b, Node *x, int k) {
    int t = b->t;
    int i = lower_bound(x->keys, x->nkeys, k);
    if (i < x->nkeys && x->keys[i] == k) {
        if (x->leaf) {
            for (int j = i; j < x->nkeys - 1; j++) x->keys[j] = x->keys[j + 1];
            x->nkeys--;
        } else if (x->children[i]->nkeys >= t) {
            int pred = subtree_max(x->children[i]);
            x->keys[i] = pred;
            btree_delete_node(b, x->children[i], pred);
        } else if (x->children[i + 1]->nkeys >= t) {
            int succ = subtree_min(x->children[i + 1]);
            x->keys[i] = succ;
            btree_delete_node(b, x->children[i + 1], succ);
        } else {
            merge_children(b, x, i);
            btree_delete_node(b, x->children[i], k);
        }
        return;
    }
    int child = i;
    if (x->children[child]->nkeys == t - 1) {
        if (child > 0 && x->children[child - 1]->nkeys >= t) borrow_prev(x, child);
        else if (child < x->nkeys && x->children[child + 1]->nkeys >= t) borrow_next(x, child);
        else if (child < x->nkeys) merge_children(b, x, child);
        else { merge_children(b, x, child - 1); child--; }
    }
    btree_delete_node(b, x->children[child], k);
}

static bool btree_delete(BTree *b, int k) {
    if (!btree_contains(b, k)) return false;
    btree_delete_node(b, b->root, k);
    if (b->root->nkeys == 0 && !b->root->leaf) {
        Node *old = b->root;
        b->root = b->root->children[0];
        free(old);
    }
    b->size--;
    return true;
}

static void btree_inorder(const Node *x, int *out, int *n) {
    for (int i = 0; i < x->nkeys; i++) {
        if (!x->leaf) btree_inorder(x->children[i], out, n);
        out[(*n)++] = x->keys[i];
    }
    if (!x->leaf) btree_inorder(x->children[x->nkeys], out, n);
}
// snippet:btree:end

static void print_ints(const char *label, const int *a, int n) {
    printf("%s: ", label);
    for (int i = 0; i < n; i++) printf("%d%s", a[i], i + 1 < n ? "," : "\n");
}

int main(void) {
    BTree b;
    btree_init(&b, 2);
    int seed[] = {10, 20, 5, 6, 12, 30, 7, 17};
    for (size_t i = 0; i < sizeof seed / sizeof seed[0]; i++) btree_insert(&b, seed[i]);

    int out[16], n = 0;
    btree_inorder(b.root, out, &n);
    print_ints("inorder", out, n);
    print_ints("root", b.root->keys, b.root->nkeys);

    printf("children: ");
    for (int i = 0; i <= b.root->nkeys; i++) {
        for (int j = 0; j < b.root->children[i]->nkeys; j++) {
            printf("%d%s", b.root->children[i]->keys[j],
                   j + 1 < b.root->children[i]->nkeys ? "," : "");
        }
        if (i < b.root->nkeys) printf(";");
    }
    printf("\n");

    int del[] = {6, 7, 5, 10, 12, 17, 20, 30};
    for (size_t i = 0; i < sizeof del / sizeof del[0]; i++) btree_delete(&b, del[i]);
    printf("size: %d\n", b.size);
    return 0;
}
